scripts/bb_regression.py

```diff
@@ -71,58 +106,118 @@ def main():
 
     device = resolve_device(opts.device)
     model = BoundingBoxRegressionModel(opts.model).to(device)
-    optimizer = th.optim.Adam(model.parameters(), lr=1e-3)
-
-    train_loader, test_loader = load_data(opts, device=device)
-
-    callbacks = Callbacks([])
-
-    # NOTE(Jiyong): for MultiBin method
-    # conf_loss_func = nn.CrossEntropyLoss().to(device)
-    # orient_loss_func = orientation_loss
-    orientation_loss = nn.MSELoss().to(device)
-    scale_loss_func = nn.MSELoss().to(device)
-
-    def loss_fn(model: th.nn.Module, data):
+    optimizer = th.optim.Adam(model.parameters(), lr=1e-5)
+    writer = SummaryWriter(path.log)
+
+    transform = Compose([CropObject(CropObject.Settings()),
+                         Normalize(Normalize.Settings(keys=(Schema.CROPPED_IMAGE,))),
+                         InstancePadding(opts.padding)])
+    train_loader, test_loader = get_loaders(opts.dataset,
+                                            device=th.device('cpu'),
+                                            batch_size=opts.batch_size,
+                                            transform=transform)
+
+    # NOTE(ycho): Synchronous event hub.
+    hub = Hub()
+
+    # Save meta-parameters.
+    def _save_params():
+        opts.save(path.dir / 'opts.yaml')
+    hub.subscribe(Topic.TRAIN_BEGIN, _save_params)
+
+    # Periodically log training statistics.
+    # FIXME(ycho): hardcoded logging period.
+    # NOTE(ycho): Currently only plots `loss`.
+    collect = Collect(hub, Topic.METRICS, [])
+    train_logger = TrainLogger(hub, writer, opts.log_period)
+
+    # Periodically save model, per epoch.
+    # TODO(ycho): Consider folding this callback inside Trainer().
+    hub.subscribe(
+        Topic.EPOCH,
+        lambda epoch: Saver(
+            model,
+            optimizer).save(
+            path.ckpt /
+            F'epoch-{epoch}.zip'))
+
+    # Periodically save model, per N training steps.
+    # TODO(ycho): Consider folding this callback inside Trainer()
+    # and adding {save_period} args to Trainer instead.
+    hub.subscribe(
+        Topic.STEP,
+        Periodic(opts.save_period, lambda step: Saver(
+            model,
+            optimizer).save(
+            path.ckpt /
+            F'step-{step}.zip')))
+
+    # Periodically evaluate model, per N training steps.
+    # NOTE(ycho): Load and process test data ...
+    # TODO(ycho): Consider folding this callback inside Trainer()
+    # and adding {test_loader, eval_fn} args to Trainer instead.
+    def _eval_fn(model, data):
+        crop_img = data[Schema.CROPPED_IMAGE].view(-1, 3, 224, 224)
```

Comment:
Hardcoded `(224,224)` value should come from settings somewhere

Reply:
Probably solved the `collate_fn` part, I expect to avoid having to use view(). But I will leave it as a comment.